### PR TITLE
Fixed bugs on the Describe> Graphs> Barchart dialog

### LINKED ISSUE #8288: Improve and fix bugs on the bar chart dialogue
@COLIEWO is working on this topic, and so this issue may partly be dealt with elsewhere.  @derekagorhom I hope you can start on these points, and @COLIEWO will be the main tester.  Perhaps @WezzieBanda also

Here is the current bar chart dialogue:
image

a) Change the title to just `Bar Charts`.  (We don't give much importance to Pie Charts, so let's not keep it in the title.)
b) None of the 3 controls `Reorder Frequency, Reorder and Position` remember their settings when you re-open the dialogue.  They should all remember.  @lloyddewit Derek mentioned that you may be doing something to fix this?  I now note that this control was ok in Version 0.7.6.  I wonder what has happened? 
c) The position of the 3 addutional `Add Labels` controls is messy.  And they are incomplere in Russian - at least.  I suggest moving all 3 to be on the same horizontally, with `Position `moved to perhaps start just below the A of Add Labels.  Maybe the same horizontally as `Colour and Size`, so moving the Add Labels checkbox up slightly.  Check that each label is wide enough that is shows completely in Russian, and maybe also check the French also.

@COLIEWO is likely to be adding to this issue as we explore the dialogue more fully

### PR DESCRIPTION
**Pull request by EstherNjeriLiberatta**
Fixes #8288 

In this PR I have made some improvements in the bar chart dialogue: 

- Changed the R-code for the pie and donut charts, when there is just a single variable - so no by factor. In that case, it "quietly" makes the constant x-variable and turns the single variable into the by.

Fixed bugs in:

- Plot options not including the geom's aesthetics - the x and y.
- Separated geoms for all the Plot options.
- The` start `checkbox in the Treemap button blocking the `ucrInputStart`.
- Still on Treemap, the checkboxes` start `and` layout` forgetting when checked when you close and reopen the dialogue.

Question:
@rdstern I noticed the `Point Options` sub-dialogue was not linked to any control while I was looking at the code first, although I linked it to the lollipop just to test it, do you still need it in the plot options?

@rdstern @N-thony @lilyclements This PR is ready for review.

### CODE REVIEW

**Review by rdstern — COMMENTED** [review]
@EstherNjeriLiberatta this is already looking great.  Well done.

First let me motivate you a bit, and you might like to discuss this general idea with @lilyclements .  The simple bar chart dialog is for a factor column.  We are wanting to show exciting things that are obvious transformations with factors - and that are often different to numeric columns.

And that is specially important in agricultural research, because the general ideas of factors can then be extended to look at factorial treatment structure in experiments. - and we like that!

I used the usual survey data.  
a) First the one variable.  It works fine.  Now try it just as a bar chart and also try the options when you click on reorder frequency and reverse, etc.  That works with bar charts, but not yet with the pie charts.
b) I think the reorder also doesn't work on the second factor yet?

c) Then I found a bug on the bar charts themselves.  I tried first with just village.  Then I used the second factor and found I had to use fill to get sensible pie or donut charts - I am happy that the user has to know that.  It gives interesting circular charts with stack, etc, and that's fine.
Then I cleared the second factor and did a bar chart again.  It still kept some of the (fill) code from the previous times, then I had 2 variables.

d) By the way I also tried pie charts with added labels and that worked fine, but I haven't tested that part very much.  That added geom_text into the code.  But we don't yet have Text Options in the pull-down for the plot options.  Would you be able to add that?

**Review by rdstern — COMMENTED** [review]
@EstherNjeriLiberatta I tried again with pie and donut, both just one variable (village) and then adding variety as the second factor.  Still using the Frequency radio button at the the top.  When I change the settings of the factor to say ascending, descending, etc, the bar chart changes, but the pie and donut don't seem to change.  I also can't see any difference in the code.
When I have the 2 factors it sometimes resets and forgets the second factor in the dialog.

I am looking forward to this all working.  Together with the discrete scales it will be a big advance.

**Comment by rdstern** [comment]
@EstherNjeriLiberatta currently the new version does not work as well as the original one for the 2 factor case.

The original needed changes when just one variable was gived, e.g. just village.  But it all was working fine for the 2 variable case, and an example of the dialog is below:

image

So the code for the 2 variables should revert to the previous one.  I hope that is easy.

For the single variable, e.g. pic chart or donut chart it was also working fine, if you included the "trick" to make it into a 2-variable case.  So add a constant column (e.g. value 1 each row) and use that as the first variable.

The changes are then to make the single variable case work, without needing the trick.  So the reorder frequency control then would work for the pie and donut.

That would be good to implement and merge.

What I then found was poor was the adding of labels.  I note that @COLIEWO made some suggestions for the treemap and wonder whether (one those are implemented) they will also imporve the labels for a pie and a donut chart.

@EstherNjeriLiberatta this is a set I am keen to merge as soon as you have the time to work on it.

**Comment by EstherNjeriLiberatta** [comment]
Working on it.

**Comment by N-thony** [comment]
> Working on it.

@EstherNjeriLiberatta we can have a chat once you are ready.

**Comment by rdstern** [comment]
@EstherNjeriLiberatta how is it going?

**Comment by EstherNjeriLiberatta** [comment]
@rdstern @N-thony This is ready for review.

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta it is close, though I am surprised you have not discovered a problem here.
I tried with the usual survey data and the one-variable now seems to be ok.

When I tried with 2 variables as so:

image

The result now ignores the village and gives a meaningless graph of the Variety!

image

The solution is easy.  The code you generate is roughly as follows:

```
# Dialog: Bar Chart
survey <- data_book$get_data_frame(data_name="survey")
last_graph <- ggplot2::ggplot(survey, ggplot2::aes(fill=variety, x=variety)) + ggplot2::geom_bar(position="fill", stat="count") + ggplot2::coord_polar(theta="y") + theme_grey()
data_book$add_object(data_name="survey", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="survey", object_name="last_graph", as_file=TRUE)

```
You can see you now have just variety as the fill as well as the x in the formula above.  You have "lost" the village.  If I make the fill = village it seems to work fine.

**Comment by rdstern** [comment]
@EstherNjeriLiberatta I have looked again at the released version (Version 4.16) and also your pull request.
The released version is very flaky and gives errors easily.  So, it is important that your version is ready for the neext release.  I think there are some imporvements, but it has also taken some steps backwards - and I hope they are easy to fix.
Here is one example:

image

You see, for our usual survey data, we are asking for a simple bar chart of village by variety.  The result is just for variety.  When I look at the R code we used to say `ggplot2::aes(x=village, fill=variety)) `  and that was fine.  Your new code says `ggplot2::aes(fill=variety, x=variety)) `.  I am not sure quite why you have swapped over the fill and the x, though it does no harm.  But the x should be the x, i.e. village, and not variety.  I think the edit you have been trying for the pie chart option, has led you astray here!

The released version also has a big problem, in that when you use the bar chart options - or any of the other special options, they are now empty.  I think you have fixed that - which is great.  I hope it is fixed generally, i.e. for all the speical options on this dialog.  Perhaps you can check.

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta that's great.  The pie chart now seems to work fine.
The (I hope only one) item remianing is that when you return to the dialog it gives a developer error.
This has come from a regression in the dialog, I assume, when the Plot options control was changed, and it was not noticed then.

You can see the problem, if you go to any of the direct geoms in the plot options.  They are blank, so the x and sometimes y, are not completed.  It should be easy to fix, but you may need help from someone who understands the system better.  So ask - maybe @anastasia-mbithe or @Vitalis95 or @N-thony, if you are not sure what to do.  It is a problem to fix as simply as possible and not to struggle over.

**Comment by EstherNjeriLiberatta** [comment]
@N-thony I updated this branch yesterday I hope that doesn't cause any issues.

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta this is looking great.  I have just been using it simply, but:
a) There is no longer the developer error when re-opening the dialog.  At least not so far.
b) I also used to get an error when changing data frames, e.g. from diamonds to survey.  This now seems ok.
c) The pie chart works fine for one variable as well as for two.
This is a great relief.  Well done.

Just one problem so far.
Use survey with village, then also variety (say).  Change the option to Fill.  This works fine.
Now get rid of the variety, so just a simple bar chart of village.  It still gives the fill.  Perhaps it also still gives other options, but that is the one I tested.  It should have cleared that argument.

Now another "interesting feature or two with the simple pie chart.

I can change the order of the bars to ascending, descending, reverse.  For the bar chart it works fine, but when I have the pie (or I presume also donut options, then it changes more:

image

And when I add labels I get a more interesting option, I quite like, namely here:

image




@lloyddewit I'd like to get this merged as soon as we can.  So happy if you could start examining the code while the last tweaks are being made.

**Comment by EstherNjeriLiberatta** [comment]
@Vitalis95 Kindly assist me on reviewing this.

File: instat/dlgBarAndPieChart.vb [hunk 1]
```diff
@@ -455,13 +461,19 @@ Public Class dlgBarAndPieChart
...
         clsBarAesFunction.SetPackageName("ggplot2")
         clsBarAesFunction.SetRCommand("aes")
 
+        clsPieFunction.SetPackageName("ggplot2")
+        clsPieFunction.SetRCommand("aes")
+        clsPieFunction.AddParameter("x", Chr(34) & Chr(34))
+        clsPieFunction.AddParameter("y", Chr(34) & Chr(34))
+
         clsPieAesFunction.SetPackageName("ggplot2")
         clsPieAesFunction.SetRCommand("aes")
```
[the remark this answers is not in the document]
Comment (EstherNjeriLiberatta) [reply instat/dlgBarAndPieChart.vb:476]: Thank you

**Comment by lloyddewit** [comment]
> @lloyddewit I'd like to get this merged as soon as we can. So happy if you could start examining the code while the last tweaks are being made.

@rdstern I looked at the code changes. I should be able to review this quite quickly after your test/approval.

**Review by rdstern — APPROVED** [review]
@EstherNjeriLiberatta and @N-thony it keeps improving!  I am sure the code can still be improved, because I have to be careful on the order of doing anything. 

image

I hope you are able and have the time to make it more solid.  But it is so much better than the existing version, that I an tentatively accepting it this way anyway.  It just seems to be what isn't cleared in the earleir R-code, as you try different options.

File: instat/dlgBarAndPieChart.vb (hunk 1 above)
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgBarAndPieChart.vb:476]: @N-thony I'm not sure I understand this comment. To me, lines 470-471 and 475-476 look different. Am I missing something?

File: instat/dlgBarAndPieChart.vb [hunk 2]
```diff
@@ -592,19 +608,17 @@ Public Class dlgBarAndPieChart
...
-        ucrReceiverByFactor.AddAdditionalCodeParameterPair(clsReorderFunctionValue, New RParameter("x", 0), iAdditionalPairNo:=5)
+        ucrReceiverByFactor.AddAdditionalCodeParameterPair(clsIsEqualToOperator1, New RParameter("left", 0), iAdditionalPairNo:=1)
+        ucrReceiverByFactor.AddAdditionalCodeParameterPair(clsIsEqualToOperator2, New RParameter("left", 0), iAdditionalPairNo:=2)
+        ucrReceiverByFactor.AddAdditionalCodeParameterPair(clsLevelsFunction, New RParameter("x", 0), iAdditionalPairNo:=3)
+        ucrReceiverByFactor.AddAdditionalCodeParameterPair(clsReorderFunctionValue, New RParameter("x", 0), iAdditionalPairNo:=4)
         ucrReceiverX.AddAdditionalCodeParameterPair(clsReorderFunction, New RParameter("x", 0), iAdditionalPairNo:=1)
-        ucrReceiverFill.AddAdditionalCodeParameterPair(clsGeomTreemapTextAesFunction, New RParameter("label", 0), iAdditionalPairNo:=1)
+        'ucrReceiverFill.AddAdditionalCodeParameterPair(clsGeomTreemapTextAesFunction, New RParameter("label", 1), iAdditionalPairNo:=1)
+        'ucrReceiverArea.AddAdditionalCodeParameterPair(clsGeomTreemapTextAesFunction, New RParameter("area", 0), iAdditionalPairNo:=1)
```
Comment (lloyddewit) [inline instat/dlgBarAndPieChart.vb:617]:
```suggestion
```

File: instat/dlgBarAndPieChart.vb [hunk 3]
```diff
@@ -870,41 +936,71 @@ Public Class dlgBarAndPieChart
...
-            If ucrChkBacktoback.Checked Then
-                clsBaseOperator.RemoveParameterByName("geom_bar")
-            Else
-                clsBaseOperator.AddParameter("geom_bar", clsRFunctionParameter:=clsRgeomBarFunction, iPosition:=2)
-            End If
+            'If ucrChkBacktoback.Checked Then
+            '    clsBaseOperator.AddParameter("geom_bar", clsRFunctionParameter:=clsRgeomBarFunction, iPosition:=2)
+            'Else
+            '    clsBaseOperator.RemoveParameterByName("geom_bar")
+            'End If
```
Comment (lloyddewit) [inline instat/dlgBarAndPieChart.vb:950]:
```suggestion
```

**Comment by EstherNjeriLiberatta** [comment]
@lloyddewit Thank you for the review. I have made the changes you and @N-thony suggested. There were two functions trying to do the same thing at @N-thony's comment, therefore, I had to get rid of one.
Thank you again!

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern There have been some small changes. If you can retest/approve, then we can merge, thanks

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta and @N-thony I am really pleased you are close to getting this dialog sorted out.  The code is still flaky, as you know.  I am testing with the simple survey data set.  Here are some edits to start you off.
a) You can do a bar chart where the variable is numeric - so I tried with fert.  This is fine for the bar chart, but not for the pie chart.  Here is the code:

`last_graph <- ggplot2::ggplot(survey, ggplot2::aes(x=1, fill=as.factor(fert))) + ggplot2::geom_bar(stat="count") + ggplot2::coord_polar(theta="y") + theme_grey()`

I have added the correction, that you should also include, namely you swap the parameter round here and make the fill=fert.  It needs the `add.factor( )` to be included here.

b) If you try with still just the one factor (anything now - village or fert - then the bar chart is ok, but the pie chart is not.  

This is what you currently generate:

`last_graph <- ggplot2::ggplot(survey, ggplot2::aes(x=forcats::fct_rev(f=forcats::fct_infreq(f=as.factor(fert))), fill=fert)) + ggplot2::geom_bar(stat="count") + ggplot2::coord_polar(theta="y") + theme_grey()`

That's the old code, so not changed as above.  What you should have is something like:

`last_graph <- ggplot2::ggplot(survey, ggplot2::aes(fill=forcats::fct_rev(f=forcats::fct_infreq(f=as.factor(fert))), x=1)) + ggplot2::geom_bar(stat="count") + ggplot2::coord_polar(theta="y") + theme_grey()`

Again, as in a) It is the fill you need to change, and you still need the` x=1`

I hope you can improve those cases to start with.  There are more oddities to try, when you include the `by factor `control.  I would be delighted if you correct more, or decide to make the correctrions step-by-step.

**Comment by EstherNjeriLiberatta** [comment]
@rdstern I made the changes as requested with assistance from Antoine. Thank you for correcting my code.
 
The code now looks like below when by factor is not filled and when reorder is checked for the pie chart:
`last_graph <- ggplot2::ggplot(survey, ggplot2::aes(fill=forcats::fct_rev(f=forcats::fct_infreq(f=as.factor(fert))), x=1)) + ggplot2::geom_bar(stat="count") + ggplot2::coord_polar(theta="y") + theme_grey()`

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta and @N-thony I am using the original survey data and so far I am trying the one variable bar chart and then pie and donut.  Done lots and so far, so good.  Well done.
Now tried with the 2-variable.  So far so good.  If I am picky, then with 2 variables it is ignoring a change (ascending/descing) in the first variable.  

a) Found the first problem.  I now went back to one variable.  Here is the dialog:

image

And here is the result:

image

So when it returns to one variable it is not clearing from the 2 variables.  So the fill is still there.

[survey with summaries.zip](https://github.com/africanmathsinitiative/R-Instat/files/12466417/survey.with.summaries.zip)

Now I have moved to the second option of the bars.  This is when you use the identity link and you need to adjust the code in the same way!  I hope that will now be easy?

image

Here is the one variable situation, which (again) works fine with the bar chart.  I have attached the data, already summarised for you to use as a check.

You have to make the same changes here as you did above.

Same for 2 variables - it is ok, for the simple pie charts/donut charts, but you need to adapt the code here too, when you include the ascending and desceing options.

I hope this will be much easier as you have already just about done it for the first tab?  You may need help, or to get @N-thony perhaps to attack first.  Up to you.

**Comment by EstherNjeriLiberatta** [comment]
@N-thony assisted me in fixing the issue with the reorder there was also a bug in the ` barchart plot options ` subdialogue which we also fixed. Thank you. @rdstern This is ready for review

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta so so good.  And almost there.  I tried lots of combinations with the frequencies and found just one problem - at the last minute too. 
I did two variables, with Village and Variety, etc.  and all worked well.  Then I took the Variety away, and the Fill, from the 2 variables is still there!  I do hope you can fix that, without disrupting any of the other options.

image

Now to check the second half - oh

I think that may be working for the 2 variable case, so with the village by variety combination.  So it would just need the trick in the one variable case.

Here is an example of the dialog:
image

Here is the code currently:

# Dialog: Bar Chart

```
survey_by_variety <- data_book$get_data_frame(data_name="survey_by_variety")
last_graph <- ggplot2::ggplot(survey_by_variety, mapping=ggplot2::aes(x=variety, y=count_non_missing_yield, fill="")) + ggplot2::geom_bar(stat="identity") + ggplot2::coord_polar(theta="y") + theme_grey()
data_book$add_object(data_name="survey_by_variety", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="survey_by_variety", object_name="last_graph", as_file=TRUE)
rm(list=c("last_graph", "survey_by_variety"))

```
As in the other case, you want fill=variety and x=1 I think.  I'll try:

# Dialog: Bar Chart

```
survey_by_variety <- data_book$get_data_frame(data_name="survey_by_variety")
last_graph <- ggplot2::ggplot(survey_by_variety, mapping=ggplot2::aes(x=1, y=count_non_missing_yield, fill=variety)) + ggplot2::geom_bar(stat="identity") + ggplot2::coord_polar(theta="y") + theme_grey()
data_book$add_object(data_name="survey_by_variety", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="survey_by_variety", object_name="last_graph", as_file=TRUE)
rm(list=c("last_graph", "survey_by_variety"))
```

This gives:

image

Please check that it also works with the ascending, etc.

Nearly there!

**Comment by EstherNjeriLiberatta** [comment]
@rdstern This is ready for review.

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta it will be good to mention what you think you have fixed.  This is still here and I keep reporting it:

image

This only happends after I have tried the 2 variables and then drop the second.  The fill parameter, that was added for the 2 variables never seems to get dropped.

Aha, but the one variable Value option now seems to work.  So perhaps the item above is all that remains on this dialog - at least for the bar and pie chart options.

If you are not totally fed up with this dialog, and as you have @N-thony on this as well, could you either try to get this merged then, or please have a look at the lollipop options before.  They seem not to work and I think that is because the sub-dialog x and y are not being complerted, and you fixed that for some of the other geoms,  It might be easy to fix.

**Review by rdstern — CHANGES_REQUESTED** [review]
@EstherNjeriLiberatta and @N-thony this is still here:

image

This comes when I have done two variables, so village and variety, and a pie chart, so I use fill.  Then I delete the second variable.  Why is it so difficult to delete the fill here, when we return to just one variable:

The line above is `+ ggplot2::geom_bar(stat="count") `  when I delete the fill - and then it works fine.  Please, please, see if you can arrange for this to be done in the code!

I now feel I am getting obsessed by this feature.  Notice, below, I have moved to the Value option and different data frame.  But the damn fill parameter is still there!!!

image

Let me check for other parts once this is fixed.

**Comment by EstherNjeriLiberatta** [comment]
@rdstern Apologies, that branch was not ready for testing yet. We have managed to fix the issue. Thank you for your patience.

**Review by rdstern — APPROVED** [review]
Great.  I am approving.  @lloyddewit over to you.

**Review by lloyddewit — APPROVED** [review]

**Pull request merged**